keps/sig-api-machinery/3037-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,590 @@
+- Provide a lot of nines of availability.
+- Work outside of the cluster network.
+- Provide a client-side load balancing solution to distribute the load between the available API servers.
+
+## Proposal
+
+The proposal is to extend client-go to allow to connect to multiple control-plane nodes, and for API servers to announce the list of available servers to the clients.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Kubernetes admin I want to deploy my cluster in HA without having to use external load balancers for the cluster communication.
+
+#### Story 2
+
+As a Kubernetes developer I'd like to be able to develop HA applications and controllers without requiring external components.
+
+#### Story 3
+
+As a Kubernetes architect I'd like the cluster components to be completely autonomous.
+
+### Notes/Constraints/Caveats (Optional)
+
+Network load balancing is complicated, there are many edge cases and there is no universal solution, the requirements also change and evolve with the time (see HTTP1.1 vs HTTP2).
+
+Client-go is the library used by all the cluster components to communicate with the API servers, any change or regression there can have a great impact on the project.
+
+The implementation of this feature should not modify any of current behaviors and be completely pluggable, so users can opt-in safely and the project can be sure there will not create any regression.
+
+### Risks and Mitigations
+
+The feature will be opt-in, users can disable it and keep having the same behavior.
+
+To avoid possible security issues, this feature will require HTTP2 and HTTPS in order to work.
+
+The alternative services IPs exposed are the same published in the Endpoints for the `kubernetes.default` services, the Alt-Svc headers will not be present on requests
+that are not authenticated.
+
+## Design Details
+
+There are two problems we have to solve:
+
+- get a list of available control-plane nodes
+- client capability to be able to connect to multiple control-planes
+
+### List of available control-plane nodes
+
+This problem can be solved in two different ways:
+
+1. Configuring on the client the list of available control-plane nodes
+
+This can be done extending the client-go configuration with an additional field that allows users to include a list of alternative control-plane nodes that the client can use.
+
+`k8s.io/client-go/rest/config.go`
+```go
+type Config struct {
+	// Host must be a host string, a host:port pair, or a URL to the base of the apiserver.
+	// If a URL is given then the (optional) Path of that URL represents a prefix that must
+	// be appended to all request URIs used to access the apiserver. This allows a frontend
+	// proxy to easily relocate all of the apiserver endpoints.
+	Host string
+  // AlternativeHosts must be a comma separated list if hosts, host:port pairs or URLs to the base of
+  // different apiservers. The client can use any of them to access the apiserver.
+	AlternativeHosts string
```

Comment:
IMO this is kind of a non starter since to actually use this in your cluster you have to modify kubeconfig files all over the place, some of which are generated etc.

Reply:
is how kubelet and kube-proxy works today , and so far how most components integrate, this kubeconfig change is backwards compatible, of course it needs work and change in current tooling

Reply:
My complaint is not lack of compatibility, but the opposite: it is too much work for a cluster admin to use the feature if they have to modify a ton of kubeconfig files.

Reply:
> My complaint is not lack of compatibility, but the opposite: it is too much work for a cluster admin to use the feature if they have to modify a ton of kubeconfig files.

It is an option, not a requirement. Alt-Svc is another option, which requires no modification of kubeconfig. In my case, where I am managing small number of on-premise clusters,  `AlternativeHosts` will suffice.

Reply:
Well, that was the requested.feature in the issue that originated the kep, my assumption is that people and installers are willing to do that sacrifice, but I can't backup that